dqvm/src/addresses.lisp

```diff
@@ -0,0 +1,284 @@
+    :reader total-number-of-addresses
+    :type a:non-negative-fixnum
+    :documentation "Length of the wavefunction")
+   (rank
+    :reader rank
+    :initarg :rank
+    :initform (mpi-comm-rank)
+    :documentation "MPI rank in charge of this instance.")
+   (number-of-processes
+    :accessor number-of-processes
+    :initarg :number-of-processes
+    :initform (mpi-comm-size)
+    :documentation "Number of parallel MPI ranks.")
+   (block-size
+    :reader block-size
+    :initarg :block-size
+    :type a:non-negative-fixnum
+    :initform *default-block-size*
+    :documentation "Minimum number of amplitudes per single block matrix-vector multiplication. This must be an even number and is typically chosen to be equal to two times the arity of the instruction with the maximum arity in the program to be executed.")
+   (blocks-per-process
+    :reader blocks-per-process
+    :type a:non-negative-fixnum
+    :documentation "Regular blocks per process (i.e., without including a possibly remaining block).")
+   (remainder-blocks
+    :reader remainder-blocks
+    :type a:non-negative-fixnum
+    :documentation "Total number of remainder blocks.")
+   (total-blocks
+    :reader get-total-blocks            ; XXX remove the get- prefix
+    :type a:non-negative-fixnum
+    :documentation "Total number of blocks among all address tables.")
+   (permutation
+    :reader permutation
+    :writer update-permutation
+    :initarg :permutation
+    :initform nil
+    :type permutation
+    :documentation "Last qubit permutation evaluated, stored in a format suitable for use by APPLY-QUBIT-PERMUTATION."))
+
+  (:documentation "Table of addresses handled by a single rank of a distributed QVM. The addresses within the table are computed on demand, rather than being explicitly stored."))
+
+(defmethod initialize-instance :after ((addresses addresses) &key)
+  (with-slots (number-of-qubits
+               total-number-of-addresses number-of-addresses
+               rank number-of-processes
+               block-size blocks-per-process remainder-blocks
+               total-blocks)
+      addresses
+    (assert (not (minusp number-of-qubits)))
+    (assert (not (minusp rank)))
+    (assert (plusp number-of-processes))
+    (assert (< rank number-of-processes))
+    (assert (evenp block-size))
+
+    (multiple-value-bind (bpp rem) (%blocks-per-process addresses)
+      (setf total-number-of-addresses (expt 2 number-of-qubits)
+            number-of-addresses (%number-of-addresses addresses)
+            total-blocks (%get-total-blocks addresses)
+            blocks-per-process bpp
+            remainder-blocks rem))))
+
+(defun make-addresses-like (addresses new-rank &optional (new-permutation nil new-permutation-p))
+  "Create new address table like ADDRESSES. Set the rank (and optionally the permutation) of the new instance to be NEW-RANK (and, respectively, NEW-PERMUTATION)."
+  (with-slots (number-of-processes number-of-qubits block-size permutation)
```

Comment:
Could you please elaborate on this point?

Reply:
Instead of grabbing the slots with WITH-SLOTS, it's preferable to use the reader methods, unless you need efficiency or something like that. WITH-SLOTS, to me, is a very low-level construct